[jaws.bar] 방탈출 예약 관리 1, 2 단계

park0503 opened the pull request:
## 구현 내용
- [1단계 요구 사항](https://edu.nextstep.camp/s/GiKTqpMP/ls/VxCgXrqM)에 맞춰 자체 메모리에 저장되는 데이터를 이용해 방탈출 예약 관리 시스템 웹을 구현했습니다.
- [2단계 요구 사항](https://edu.nextstep.camp/s/GiKTqpMP/ls/UKMwWkL9)를 구현했습니다.
  - 스켈레톤 코드로 구현되어 있던 콘솔 프로그램에 DB를 연결했습니다.
    - 수동으로 connection 연결
  - 1단계에서 구현했던 웹 어플리케이션에 DB를 연동했습니다.
    - jdbcTemplate 사용
  - Global Exception Handler 를 이용해 전역적으로 예외 처리를 해 주었습니다.
## 궁금점
- ~~JDBC template을 이용해 예약 생성 쿼리 메소드를 작성할 때, auto increment된 id를 반환해 주기 위해 결국 KeyHolder를 이용했는데, 이 때 결국 preparedStatement 를 이용해 주어 코드가 굉장히 더러워졌습니다. 현업에서 jdbcTemplate을 직접 이용하는 일이 많지는 않겠지만, 혹시 이 때 간단하게 자동 생성된 id를 받아올 수 있는 방법이 있는지 궁금합니다.~~
  - [다음 링크](https://docs.spring.io/spring-framework/docs/current/reference/html/data-access.html#jdbc-simple-jdbc-insert-1)를 참고해 SimpleJdbcInsert를 도입하고 SqlParameterSource를 이용해 코드를 다듬어 해결했습니다!

### 감사합니다.

baker-github committed:
feat: 예약 생성 api 구현

baker-github committed:
feat: 예약 조회 api 구현

baker-github committed:
예약 삭제 api 구현

baker-github committed:
feat: 예외처리

- Global Exception Hanlder 구현
- 중복된 시간에 예약 시도 시 예외 발생
- 존재하지 않는 예약 조회 시도 시 예외 발생

baker-github committed:
feat: 콘솔 어플리케이션 데이터베이스 적용

baker-github committed:
fix:예약 생성시 id 출력 수정

baker-github committed:
refactor: try wit resources 적용

baker-github committed:
refactor: 예약 생성 jdbc로 전환

baker-github committed:
refactor: findById, findByDateAndTime JDBC로 전환

baker-github committed:
refactor: delete JDBC로 전환

park0503 committed:
style: 미사용 코드 주석 처리

park0503 committed:
style: 디버깅용 프린트문 제거

park0503 committed:
refactor: ReservationRepository 추상화

park0503 committed:
feat: 존재하지 않는 API 호출 예외처리

park0503 committed:
feat: api 요청 시 파라미터가 잘못 되었을 때 발생하는 예외 처리

park0503 committed:
feat: 잘못된 경로 변수 및 쿼리 파라미터 예외 처리

gayeon1025 commented on src/main/java/kakao/controller/ReservationController.java:
```diff
@@ -0,0 +1,42 @@
+        this.reservationService = reservationService;
+    }
+
+    @PostMapping()
```
@RestControllersk @PostMapping에서 파라미터가 필요없을 경우 괄호는 생략 가능합니다!

park0503 replied:
헉 그렇군요! 반영하겠습니다!

gayeon1025 commented on src/main/java/kakao/domain/Reservation.java:
```diff
@@ -0,0 +1,60 @@
+    private String name;
+    private Theme theme;
+
+    public Reservation(LocalDate date, LocalTime time, String name, Theme theme) {
+        this(null, date, time, name, theme);
+    }
```
죠스~ 혹시 lombok이라는 라이브러리 들어보셨나요? 수업에서는 롬복을 사용하지 않아야한다거나 그런 가이드가 있었는지 궁금해요. 없다면 lombok은 자주 쓰이는 라이브러리라서 한 번쯤 써보는 것도 재밌을 거 같아요!

park0503 replied:
네 Lombok 라이브러리는 다른 프로젝트에서 몇 번 사용해 봤는데, 이번 과제에서는 따로 언급이 없길래 굳이 사용하지는 않았습니다. 생각해보면 굳이 사용하지 않을 이유는 없네요. 이번 과제에도 도입해 보겠습니다.

park0503 committed:
refactor: SimpleJdbcInsert 도입

park0503 committed:
feat: 기타 서버 에러 예외 처리

gayeon1025 commented on src/main/java/kakao/service/ReservationService.java:
```diff
@@ -0,0 +1,50 @@
+    public long createReservation(CreateReservationRequest request) {
+        boolean isDuplicate = reservationRepository.findByDateAndTime(request.date, request.time).size() > 0;
+        if (isDuplicate) {
+            throw new DuplicatedReservationException(ErrorCode.DUPLICATE_RESERVATION);
```
실제로도 이런식으로 에러코드 정의해서 많이 합니다 굿굿!👍

gayeon1025 commented on src/main/java/kakao/service/ReservationService.java:
```diff
@@ -0,0 +1,50 @@
+        int deletedCount = reservationRepository.delete(id);
+        if (deletedCount == 0) {
+            throw new RecordNotFoundException(ErrorCode.RESERVATION_NOT_FOUND);
+        }
```
삭제하려는 데이터가 없을 때 시스템 상에서의 멱등성은 유지되기 때문에 익셉션을 보여야하는지는 다시 고려해보는 게 좋을 것 같습니다!
[참고블로그](https://velog.io/@bae12/HTTP-%EB%A9%94%EC%86%8C%EB%93%9C%EC%9D%98-%EB%A9%B1%EB%93%B1Idempotent%EB%9E%80) 함께 첨부드립니다!

park0503 replied:
익셉션의 발생 여부로 자동 재요청을 하는 메커니즘을 이용할 수도 있군요. 단순히 정보를 알려주기 위한 용도로만 exception을 사용하려는 경향이 있는데 좀 더 확실한 사용 위치를 저 나름대로 정의내려야 할 것 같습니다. 감사합니다.

gayeon1025 commented on src/main/java/kakao/controller/ReservationController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @PostMapping()
+    public ResponseEntity<Long> createReservation(@Valid @RequestBody CreateReservationRequest request) {
+        long generatedId = reservationService.createReservation(request);
+        return ResponseEntity.created(URI.create("")).body(generatedId);
+    }
```
응답값을 Long인 이유가 있을까요? 실제 예약 사이트들에서 예약 완료 후에 예약현황을 다시 한 번 보여주는 경우가 많은데, 저장된 Reservation이 아닌 id를 리턴하면 예약 현황 페이지를 위해 id로 다시 한 번 조회가 일어나야할 것 같아요

park0503 replied:
음 사실 요구사항 링크를 보면 예약 생성 API 의 응답이 따로 없는 것으로 되어 있어서 처음에 Void로 두었다가, 2단계에서부터 콘솔 프로그램의 요구사항을 만족 시키려면 id가 필요해 다음과 같이 id만을 리턴해 주는 식으로 구현했습니다.

말씀해 주신 대로 생성 api는 생성된 객체의 정보를 return해 주는 것이 더 일반적인 flow 같습니다. 과제 요구사항에 너무 매몰되어 있었던 것 같네요. 좀 더 확장해서 받아들일 필요가 있는 것 같습니다. 감사합니다!

gayeon1025 commented on src/main/java/kakao/dto/request/CreateReservationRequest.java:
```diff
@@ -0,0 +1,29 @@
+
+public class CreateReservationRequest {
+
+    @NotNull
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
+    public final LocalDate date;
+
+    @NotNull
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "HH:mm")
+    public final LocalTime time;
```
LocalDateTime으로 통합하면 date와 time을 한 번에 받을 수 있을 것 같아요! 혹시 날짜와 시간을 분리한 이유가 있었을까요?

이건 추가적인 내용이지만 글로벌 서비스인 경우 타임존이 포함된 ZonedDateTime 등을 쓰기도 합니다![[참고](https://velog.io/@hazzang/LocalDateTime-vs-Instant-vs-ZonedDateTime)]

gayeon1025 replied:
요거 1단계 요구 사항에 나와있는 도메인을 차용하신 것 같아서 꼭 반영하지 않으셔도 될 것 같네요~

park0503 replied:
별 다른 이유는 없고 스켈레톤 코드에 정의되어 있던 Reservation 클래스가 다음과 같이 나누어져 있어서 그대로 사용했습니다. 보통 처음부터 구현할 때에는 저도 LocalDateTime을 사용합니다. 굳이 나눠져 있을 필요가 없을 것 같네요. ZonedDateTime에 관해서는 몰랐던 내용이에요. 확장성을 생각하면 일반적인 LocalDateTime보다 나을 것 같네요! 감사합니다!

park0503 committed:
feat: 예약 응답 dto date, time 포맷 설정

gayeon1025 commented on src/main/java/kakao/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,82 @@
+
+    //Custom
+
+    // 이미 예약이 되어 있는 시간대에 예약을 시도할 경우 발생
+    @ExceptionHandler(DuplicatedReservationException.class)
+    public ResponseEntity<ErrorResponse> handleDuplicatedReservationException(DuplicatedReservationException e) {
+        return handlerException(e.getErrorCode());
+    }
+
+    // 해당 id의 row 가 존재하지 않을 시 발생
+    @ExceptionHandler(RecordNotFoundException.class)
+    public ResponseEntity<ErrorResponse> handleRecordNotFoundException(RecordNotFoundException e) {
+        return handlerException(e.getErrorCode());
+    }
```
이처럼 GlobalExceptionHandler을 구성해주시면 모든 exception에 대해 의도한 ErrorResponse를 내려줄 수 있는 장점이 있을 것 같습니다! 다만 서비스가 확장되는 것을 고려해보면 기능이 추가됨에 따라 exception이 자꾸 늘어나게 될 테고 GlobalExceptionHandler가 너무 비대해질 것 같아요. 커스텀하게 정의된 exception의 경우는 항상 아래와 같은 exceptionHandler 코드가 만들어질 것 같구요
```
@ExceptionHandler(XXXException.class)
public ResponseEntity<ErrorResponse> handleXXXException(XXXException e) {
    return handlerException(e.getErrorCode());
}
```
항상 위와 같은 패턴이라면 해당 시스템에서 발생하는 커스텀 exception들을 하나의 interface를 구현하도록 하면 커스텀 익셉션은 하나의 메서드로 핸들링할 수 있을 것 같은데 어떻게 생각하시나요??

park0503 replied:
저도 이 부분을 보면서 한 줄이 계속 중복되는게 마음에 걸렸었는데, 인터페이스 생각을 못했네요! 당장 적용하겠습니다. 감사합니다.

gayeon1025 commented on src/main/java/kakao/service/ReservationService.java:
```diff
@@ -0,0 +1,50 @@
+        if (isDuplicate) {
+            throw new DuplicatedReservationException(ErrorCode.DUPLICATE_RESERVATION);
+        }
+        Reservation reservation = new Reservation(request.date, request.time, request.name, ThemeRepository.theme);
```
(ThemeRepository.theme은 아마 다음 과제에서 테마확장을 염두해주시고 작성하신 코드같네요!)

함수의 파라미터를 최대 3개까지만 두는 것이 가독성을 높이고 실수할 확률을 낮출 수 있는데요! 
생성자의 경우 생성시에 필요한 파라미터가 많다면 빌더패턴을 활용해보는 것은 어떨까요?

park0503 replied:
위에서 말씀해주신 Lombok을 이용해 깔끔하게 builder 패턴을 활용할 수 있을 것 같습니다. 감사합니다.

park0503 committed:
refactor: 피드백 반영

- annotation 뒤 미사용 괄호 제거
- lombok 라이브러리 도입
- delete api 예외 발생 제거
- create api 가 예약 정보를 반환하도록 수정

park0503 committed:
refactor: CustomException 일반화

park0503 committed:
refactor: lombok을 이용한 빌더패턴 적용

park0503 commented:
피드백 반영했습니다!

CustomException의 경우 `ExceptionHanlder` 어노테이션의 파라미터로는 throwable한 클래스밖에 넘기지 못하기에 interface 가 아닌 RuntimeException을 상속한 추상 클래스로 만들어 주었습니다.

gayeon1025 commented:
@park0503 피드백 반영하시느라고 고생 많으셨습니다! 원래 코드도 잘 작성해주셔서 별로 커멘트 드릴 게 없었네요ㅎㅎㅎ

> CustomException의 경우 ExceptionHanlder 어노테이션의 파라미터로는 throwable한 클래스밖에 넘기지 못하기에 interface 가 아닌 RuntimeException을 상속한 추상 클래스로 만들어 주었습니다.
getErrorCode는 익셉션마다 다르게 구현할 여지가 크지 않을 것 같아요. 메서드명을 보면 항상 `return errorCode`일 것 같아서 부모에 구현을 하고 CustomException이 추상클래이지 않아도 될 것 같아요. 더불어 익셉션은 항상 errorCode 속성을 갖고 있을 것 같으니 마찬가지로 부모에 선언을 해도 괜찮을 것 같네요.

gayeon1025 reviewed (approved):
죠스 고생많으셨습니다! 
최종으로 반영해주신 코드에 대해 커멘트 남겨드렸습니다! 다음 3단계에 함께 고민해주시면 좋을 것 같아요.
더불어 이번 과제에서는 테스트코드가 아직 없는 것 같아요. 이전 과제까지는 유닛테스트만 작성해보셨을텐데 이번 기회에 intergration test도 함께 작성해보면 좋을 것 같습니다! ([참고](https://www.baeldung.com/integration-testing-in-spring))
